src/exchanges/bitfinex-client.js

```diff
@@ -207,9 +234,79 @@ class BitfinexClient extends BasicClient {
   }
 
   _onTradeMessage(msg, market) {
-    // example msg: [ 359491, 'tu', [ 560287312, 1609712228656, 0.005, 33432 ], 6 ]
-    let [id, unix, amount, price] = msg[2];
+    const timestampMs = msg[3];
+    if (msg[1] === "hb") {
+      const sequenceId = Number(msg[2]);
+      if (this.enableEmptyHeartbeatEvents === false) return;
+      // handle heartbeat by emitting empty update w/sequenceId.
+      // example trade heartbeat msg: [ 198655, 'hb', 3, 1610920929093 ]
+      let trade = new Trade({
+        exchange: "Bitfinex",
+        base: market.base,
+        quote: market.quote,
+        timestamp: timestampMs,
+        sequenceId,
+      });
+      this.emit("trade", trade, market);
+      return;
+    }
+    if (Array.isArray(msg[1])) {
+      const sequenceId = Number(msg[2]);
+      // trade snapshot example msg:
+      /*
+      [
+        CHANNEL_ID,
+        [
+          [
+            ID,
+            MTS,
+            AMOUNT,
+            PRICE
+          ],
+          ...
+        ],
+        sequenceId,
+        timestampMs
+      ]
+      */
+      msg[1].forEach(thisTrade => {
+        let [id, unix, amount, price] = thisTrade;
+
+        let side = amount > 0 ? "buy" : "sell";
+        price = price.toFixed(8);
+        amount = Math.abs(amount).toFixed(8);
+        let trade = new Trade({
+          exchange: "Bitfinex",
+          base: market.base,
+          quote: market.quote,
+          tradeId: id.toFixed(),
+          sequenceId,
+          unix: unix,
+          side,
+          price,
+          amount,
+        });
+        this.emit("trade", trade, market);
+      });
+      return;
+    }
+    // example trade update msg: [ 359491, 'tu' or 'te', [ 560287312, 1609712228656, 0.005, 33432 ], 6 ]
+    // note: "tu" means it's got the tradeId, this is delayed by 1-2 seconds and includes tradeId.
+    // "te" is the same but available immediately and without the tradeId
+    let shouldHandleTradeEvent = false;
+    const tradeEventType = msg[1];
+    if (this.tradeMessageType === "all") {
```

Comment:
This complexity should be moved into the `_onMessage` method, which acts as a router